File: javascript/oop-patterns/behavioral-patterns/interpreter/interpreter.js
// терминальное выражение
class Numeral {
  value;

  constructor(value) {
    this.value = parseInt(value);
  }

  interpret() {
    return this.value;
  }
}

// Нетерминальное выражение
class Add {
  left;
  right;

  constructor(left, right) {
    this.left = left;
    this.right = right;
  }

  interpret() {
    return this.left.interpret() + this.right.interpret();
  }
}

class Subtract {
  left;
  right;

  constructor(left, right) {
    this.left = left;
    this.right = right;
  }

  interpret() {
    return this.left.interpret() - this.right.interpret();
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

// Интерпретируемая строка состоит из чисел, пробелов и знаков "+" и "-"
const SENTENCE = '5 + 4 - 3 + 7 - 2';
console.log(SENTENCE);

// В данном случае создание Абстрактного синтаксического дерева (AST) происходит
// вручную
const AST = [];
AST.push(new Add(new Numeral('5'), new Numeral('4'))); // 5 + 4
AST.push(new Subtract(AST[0], new Numeral('3'))); // ^ - 3
AST.push(new Add(AST[1], new Numeral('7'))); // ^ + 7
AST.push(new Subtract(AST[2], new Numeral('2'))); // ^ - 2

const AST_ROOT = AST.pop();

console.log(AST_ROOT.interpret()); // => 11

File: go/oop-patterns/behavioral-patterns/interpreter/interpreter.go
package main

import (
	"fmt"
	"strconv"
)

// каждое выражение (как терминальное, так и нетерминальное) должно иметь метод
// для интерпретации
type iAbstractExpression interface {
	interpret() int
}

func newNumeral(s string) *numeral {
	parsedInt, err := strconv.Atoi(s)
	if err == nil {
		return &numeral{value: parsedInt}
	}
	return nil
}

// терминальное выражение
type numeral struct {
	value int
}

func (n *numeral) interpret() int {
	return n.value
}

// Нетерминальное выражение
type add struct {
	left  iAbstractExpression
	right iAbstractExpression
}

func newAdd(left, right iAbstractExpression) *add {
	return &add{left: left, right: right}
}

func (a *add) interpret() int {
	return a.left.interpret() + a.right.interpret()
}

type subtract struct {
	left  iAbstractExpression
	right iAbstractExpression
}

func newSubtract(left, right iAbstractExpression) *subtract {
	return &subtract{left: left, right: right}
}

func (s *subtract) interpret() int {
	return s.left.interpret() - s.right.interpret()
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// ------------------------------- Клиентский код -------------------------------
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
func main() {
	// Интерпретируемая строка состоит из чисел, пробелов и знаков "+" и "-"
	SENTENCE := "5 + 4 - 3 + 7 - 2"
	fmt.Println(SENTENCE)

	// В данном случае создание Абстрактного синтаксического дерева (AST) происходит
	// вручную
	AST := []iAbstractExpression{}

	AST = append(AST, newAdd(newNumeral("5"), newNumeral("4"))) // 5 + 4
	AST = append(AST, newSubtract(AST[0], newNumeral("3")))     // ^ - 3
	AST = append(AST, newAdd(AST[1], newNumeral("7")))          // ^ + 7
	AST = append(AST, newSubtract(AST[2], newNumeral("2")))     // ^ - 2

	fmt.Println(AST[3].interpret()) // => 11

}
